Add social sharing functionality

kcne opened the pull request:
### Add social network sharing to diary entry pages

**Description:**
This PR addresses the issue #1391 by introducing social media sharing functionality for diary entries, allowing users to share posts on platforms like Twitter, Facebook, LinkedIn, Mastodon, Telegram, and Email. It includes required JavaScript to handle opening share URLs in pop-up windows and Ruby helper methods to filter and validate allowed social media sites. The changes also incorporate error logging for invalid sites and missing icons, ensuring robustness.


**Implementation details:**
- `lib/social_share_button_helper.rb` - config, validation & helper functions
- `render_social_share_buttons` function inside `application_helper.rb` responsible for rendering the buttons with helper utilization
- `social_share_button.js` - javascript utility functions for opening the popups, generating urls for sharing & handling onclick events on buttons

**Screenshots:**
<img width="209" alt="Screenshot 2024-07-17 at 16 05 18" src="https://github.com/user-attachments/assets/bcb3c286-b9d7-4d8a-9a11-ab46644001cd">
<img width="638" alt="Screenshot 2024-07-17 at 16 07 12" src="https://github.com/user-attachments/assets/fbd15fe0-31a0-4f85-8f0e-6c67dec94ffb">


Reviews welcome,
Thank you.

mmd-osm commented:
Thank you. Some of the links seem to be a bit dated, e.g. Twitter is now known as X, and links/logos would have to be updated, I suppose.

kcne commented:
Updated the x icon and link as suggested by @mmd-osm. Also if you think we should include some other socials too, suggestions are welcome. Any chance @tomhughes @gravitystorm @AntonKhorev some of you guys can review this soon?
Thank you.

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,45 @@ def render_flash(flash)
+          next
+        end
+
+        link_to "", {
```
Why do links to social media have to be generated by javascript instead of being available here?

kcne replied:
Good point, just refactored according to this

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,45 @@ def render_flash(flash)
+      Rails.logger.error("Invalid site or icon not configured: #{invalid_site}")
+    end
+
+    content_tag(
+      :div,
```
```suggestion
    tag.div(
```

kcne replied:
✔️

AntonKhorev commented on app/views/diary_entries/_diary_entry.html.erb:
```diff
@@ -1,3 +1,7 @@
+<% content_for :head do %>
+  <%= javascript_include_tag "social_share_button" %>
```
You'll have as many `<script>` tags as there are articles and a bunch of js errors about redeclarations.

kcne replied:
Added this into `_head.html.erb`

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,45 @@ def render_flash(flash)
+          :rel => ["nofollow", opts[:rel]].compact,
+          :data => { :site => site, :title => opts[:title], :desc => opts[:desc], :url => opts[:url] },
+          :class => "social-share-button ssb-icon",
+          :onclick => "return SocialShareButton.share(this);"
```
`onclick` is probably some leftover because it refers to `SocialShareButton` which doesn't exist

kcne replied:
True that. Just removed it completely.

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,45 @@ def render_flash(flash)
+        :via => opts[:via]
+      }
+    ) do
+      social_buttons = valid_sites.filter_map do |site|
```
Why assigning to `social_buttons` if you return this value right away?

kcne replied:
Just a habit thing, now it's gone.

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,45 @@ def render_flash(flash)
+          :data => { :site => site, :title => opts[:title], :desc => opts[:desc], :url => opts[:url] },
+          :class => "social-share-button ssb-icon",
+          :onclick => "return SocialShareButton.share(this);"
+        } do
```
There probably should also be a `title` attribute with "Share on (platform)" value.

kcne replied:
Makes sense, so I added it.

AntonKhorev commented on app/assets/javascripts/social_share_button.js:
```diff
@@ -0,0 +1,59 @@
+  const newWindow = window.open(url, "popup", opts);
+
+  if (!newWindow || newWindow.closed || typeof newWindow.closed == "undefined") {
+    console.error("Popup blocked. Please allow popups for this website.");
```
Who is supposed to be able to read this message? Regular users don't look at their browser console.

kcne replied:
Removed this.

talllguy commented:
Thanks for the hard work on this! Since we're on the topic of social sharing I thought I'd resurface an idea from 10 years ago that I still think would add value to users of the site. Perhaps the technology hurdles back then are less likely to be hurdles now. https://github.com/openstreetmap/openstreetmap-website/issues/739

kcne commented:
Refactored this according to @AntonKhorev review. Please let me know If you think I can further improve this. Thank you!

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,29 @@ def render_flash(flash)
+        link_to SocialShareButtonHelper.generate_share_url(site, opts), {
+          :rel => ["nofollow", opts[:rel]].compact,
+          :class => "ssb-icon",
+          :title => "Share on #{site.capitalize}"
```
This needs to be translated.

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,29 @@ def render_flash(flash)
+          :class => "ssb-icon",
+          :title => "Share on #{site.capitalize}"
+        } do
+          image_tag(SocialShareButtonHelper.icon_path(site), :alt => site.capitalize, :size => 28)
```
`:alt => site.capitalize`: "email" is not a proper site name, you'll have to translate it instead of just capitalizing. 

You'll have something like:
https://github.com/openstreetmap/openstreetmap-website/blob/1b288d40c7fff3d3be8d82d4e96c9008778f4dba/config/locales/en.yml#L2597-L2614

AntonKhorev commented on app/views/diary_entries/_diary_entry.html.erb:
```diff
@@ -44,4 +44,10 @@
       <% end %>
     </ul>
   </nav>
+
```
The buttons ate too close to the actions list above. Probably need to add margins to it.
![image](https://github.com/user-attachments/assets/e5f28f53-ca9d-4642-b463-46e2c4d87daf)

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,29 @@ def render_flash(flash)
+      valid_sites.map do |site|
+        link_to SocialShareButtonHelper.generate_share_url(site, opts), {
+          :rel => ["nofollow", opts[:rel]].compact,
+          :class => "ssb-icon",
```
If all icons are going to be circular, you can add the `rounded-circle` class to get circular focus outlines:
![image](https://github.com/user-attachments/assets/1579367a-7572-40af-a5bc-98fe40c71d1f)

AntonKhorev commented on lib/social_share_button_helper.rb:
```diff
@@ -0,0 +1,48 @@
+    site = site.to_sym
+    case site
+    when :email
+      "mailto:?subject=#{URI.encode_www_form_component(params[:title])}&body=#{URI.encode_www_form_component(params[:url])}"
```
`URI.encode_www_form_component` is a bit different from javascript's `encodeURIComponent`. It encodes spaces as `+`. But [mailto rfc](https://datatracker.ietf.org/doc/html/rfc6068#section-5) says spaces should be encoded as `%20`.

tomhughes reviewed (changes requested):
Added a few comments inline... Other things to consider are whether to offer some other sites - bluesky and threads being the obvious candidates.

kcne commented on app/assets/javascripts/social_share_button.js:
```diff
@@ -0,0 +1,27 @@
+// Opening pop-ups with share URL
+function openShareUrl(url, initialWidth = 640, initialHeight = 480) {
+  if (typeof url !== "string" || !url.startsWith("http")) {
+    console.error("Invalid URL"); // Consider removing this line if console warnings should be avoided.
```
In current state it is not really possible, since all the urls are hardcoded.

kcne commented on lib/social_share_button_helper.rb:
```diff
@@ -0,0 +1,48 @@
+    when :facebook
+      "https://www.facebook.com/sharer/sharer.php?url=#{URI.encode_www_form_component(params[:url])}"
+    when :mastodon
+      "https://mastodon.social/share?text=#{URI.encode_www_form_component(params[:title])}&url=#{URI.encode_www_form_component(params[:url])}"
```
Good catch but I'm a little stuck here, there is n numbers of mastodon instances. Do you have some recommendation on how to approach this?

kcne replied:
Maybe use [mapstodon.space](mapstodon.space)?

mmd-osm replied:
A lot of folks hang out on en.osm.town. Maybe check out https://github.com/Juerd/tootpick?tab=readme-ov-file which offers a popup to enter the mastodon instance (might be include a few better known ones as preselect?).

kcne replied:
Found out there is [mastodonshare.com](https://mastodonshare.com) that takes text and url query parameters and let's you choose the mastodon instance you want to share on. Updated the code accordingly. Can you please test this and let me know if this solution is okay. @mmd-osm @tomhughes

kcne commented:
Thank you for the review!
I've resolved the requested changes, added the translations as requested, and the mailto links should now be opening correctly. Once we round this up and resolve the mastodon thing with @tomhughes  I can add bluesky and threads support as a separate commit if you agree.

kcne commented:
> Found out there is [mastodonshare.com](https://mastodonshare.com) that takes text and url query parameters and let's you choose the mastodon instance you want to share on. Updated the code accordingly. Can you please test this and let me know if this solution is okay. @mmd-osm @tomhughes

Updated mastodon link, everything should be working fine now. Can you please check and let me know if anything else can be improved. Thank you.

AntonKhorev commented on app/assets/javascripts/social_share_button.js:
```diff
@@ -0,0 +1,27 @@
+  window.open(url, "popup", opts);
+}
+
+$(document).ready(function () {
+  $(".ssb-icon").on("click", function (e) {
```
This is not going to work with Turbo paging because `$(document).ready` won't happen when you click *Newer/Older Entries*. This can be fixed by delegating the event to something that stays on the page or the entire document, but the other question is: Should there be share buttons on multiple entry pages at all? We're also trying to truncate the entries on such pages, with #5121 or something like. You made your own version in #5125 and I don't see share buttons in it. Maybe the buttons should only be present on single entry pages?

kcne replied:
I agree on this one. While working on this, there was not too much information indicating someone is going to do such thing. If we are going to truncate the text, then this definitely should only be displayed on single entry pages. Although I'm not sure should I do it right away or wait for #5121 to be merged. Thank you.

AntonKhorev replied:
It's already not working correctly on index pages so you'll have to do something. Removing the buttons is easier than fixing them, right?

AntonKhorev commented on app/assets/javascripts/social_share_button.js:
```diff
@@ -0,0 +1,27 @@
+  const top = (screen.height * 0.3) - (height / 2);
+  const opts = `width=${width},height=${height},left=${left},top=${top},menubar=no,status=no,location=no`;
+
+  window.open(url, "popup", opts);
```
Email link only works as expected for me when it's *not* opened with `window.open`.

It works when middle-clicked, or when the event handler is broken because of Turbo + `$(document).ready`.

AntonKhorev replied:
Still opens a window for `mailto:` links. What do you expect to appear in that window?

kcne replied:
Expected behaviour is to open mailto link in new tab where you can select your mail app/provider, which should be working for the timebeing. Do you suggest we should handle mail sharing differently?

AntonKhorev replied:
But you're also opening a new window that's not going to contain anything and will have to be closed manually. Or is it working differently for you?

kcne replied:
Actually for me it works like this - I'm using firefox, hadn't check for the other browsers.

When I click share it opens share link in pop up window to select the email provider:
<img width="1213" alt="Screenshot 2024-10-22 at 10 07 59" src="https://github.com/user-attachments/assets/d7eee376-638c-4684-ae79-96efecffd0b5">
After email provider selection, i get the automatically formatted email message ready to be sent:
<img width="886" alt="Screenshot 2024-10-22 at 10 08 23" src="https://github.com/user-attachments/assets/804f34bb-2ef3-43af-a615-df2b5bebd23f">

AntonKhorev replied:
Here's how it looks for me:
![image](https://github.com/user-attachments/assets/b84a3fbc-b52a-4223-aef3-5f09a19e218d)

I get a blank window in background and a new tab.

kcne replied:
In this case the email dialog gets opened right away. Fixed now, thanks.

kcne commented:
As @AntonKhorev suggested, we're now displaying these icons only when users open the full diary entry, as outlined in #5121. Since this is currently used in one place, turbo pagination issues can be handled in a future PR if a use case arises for combining it with turbo pagination. Thank you.

AntonKhorev reviewed (approved):
It's fine if we're ok with mastodonshare.com for now, but we can replace it later.

AntonKhorev commented on app/views/diary_entries/show.html.erb:
```diff
@@ -1,3 +1,5 @@
+ <%= javascript_include_tag "social_share_button" %>
```
```suggestion
<% content_for :head do %>
  <%= javascript_include_tag "social_share_button" %>
<% end %>
```

That's how we do it, for example, in https://github.com/openstreetmap/openstreetmap-website/blob/fe81ac334cfef04751a959297dca24e472cb025f/app/views/dashboards/show.html.erb#L13-L15

kcne replied:
Thanks for the tip. Updated now.

AntonKhorev replied:
But it's not updated.

kcne replied:
Updated now, forgot to push, sorry.

AntonKhorev commented on app/views/diary_entries/_diary_entry.html.erb:
```diff
@@ -44,4 +44,5 @@
       <% end %>
     </ul>
   </nav>
+
```
We probably don't need to touch this file?

kcne replied:
You're right. I don't know how I missed this. I've restored the changes to this file and updated the commit. Thank you.

AntonKhorev commented:
Merged, thanks.

Although we're effectively outsourcing Mastodon host option to mastodonshare.com, it works for anonymous users too. Currently our preferences page is not available for anonymous users but we probably should make it available and save preferences to cookies. We have other preferences like #5201 and color mode settings and we shouldn't add a button for each one.

AntonKhorev commented on app/helpers/application_helper.rb:
```diff
@@ -75,4 +76,32 @@ def render_flash(flash)
+
+  # Generates a set of social share buttons based on the specified options.
+  def render_social_share_buttons(opts = {})
+    sites = opts.fetch(:allow_sites, [])
```
We can probably use it if sharing on some site requires some setting, to filter out that site if it's not set, like if we add Mastodon host preference.

gravitystorm commented:
> anonymous users

I think you mean users who are not logged in - "anonymous users" is the term we use for accounts with `user.data_public == false`. They are quite rare now since [they haven't been permitted for a long time](https://wiki.openstreetmap.org/wiki/Anonymous_edits).

As for preferences, I'm happy keeping most preferences available for only logged in users, since account creation is free and we aim to serve our mappers (who by definition have accounts) more than just casual visitors. Storing preferences in cookies does have drawbacks (like cross-device consistency, automatic cookie expiry etc) so it's not my first choice in general.

AntonKhorev commented:
> I think you mean users who are not logged in

yes

> As for preferences, I'm happy keeping most preferences available for only logged in users, since account creation is free and we aim to serve our mappers (who by definition have accounts) more than just casual visitors. Storing preferences in cookies does have drawbacks (like cross-device consistency, automatic cookie expiry etc) so it's not my first choice in general.

Are you against the entire idea of #1115 or just against the cookies and want preferences stored somewhere else like session? I don't think anyone who is not logged in expects cross-device consistency etc.
